### DISCUSSION TITLE
How to write to OASIS ( *.oas ) for more than 32767 columns/rows

### DISCUSSION THREAD

**Discussion by ghost** [gdsfactory/gdsfactory#3476] [Q&A] [answered]
How do I / Can I specify or indicate to the write function to output an OASIS file to write an array larger than the GDS2 limitation?

```python
import gdsfactory as gf

@gf.cell
def big_array(size:tuple=(1,1),
                layer:tuple=(1,0),
                columns:int=3,
                rows:int=3,
                column_pitch:float=3,
                row_pitch:float=3,
                ) -> gf.Component:
    c = gf.Component()
    ref1 = c.add_ref(gf.components.rectangle(size=size, layer=layer),
                    columns = columns,
                    rows = rows,
                    column_pitch = column_pitch,
                    row_pitch = row_pitch)
    return c

c = big_array(columns=32767, rows = 32768)

c.write('big_array.oas')
#c.write_gds(gdspath='big_array.oas')
```
```
RuntimeError: Cannot write array references with more than 32767 columns or rows to GDS2 streams, 
writing instances, writing cell 'big_array_S1_1_L1_0_C32_2b680d99' in Cell.write"
```

**Comment by tvt173** [comment]
i can see that it's properly writing to OASIS (for smaller array sizes), using the klayout `Cell.write()` function, and the error comes from here
https://github.com/KLayout/klayout/blob/9441024fd08805c6675de1c2b3ad4d6fa3f4f627/src/plugins/streamers/gds2/db_plugin/dbGDS2WriterBase.cc#L702

if this limit should not be checked on OASIS files, I think this probably a bug in klayout

@klayoutmatthias
@sebastian-goeldi

**Reply by sebastian-goeldi** [reply]
I am not sure. To my knowledge OASIS is not an open standard, therefore hard to know ;)

It's still very much possible it has the same limitation (both array coordinates will nicely fit into one 32bit int).

**Reply by sebastian-goeldi** [reply]
Alsoz pretty sure the error doe not come from there, it should be in the oasis writer.

**Reply by tvt173** [reply]
Well, I'm not very good with the klayout codebase, but that was the only place I was able to find that error message...

**Comment by ghost** [comment]
I did make a similar array in klayout, but I do not know how the two different calls differ.

```python
import klayout.db as db

ly = db.Layout()
f_cell = ly.create_cell("F")
l1 = ly.layer(1, 0)

point_cell = ly.create_cell('P')
pt = [[0,0]]
point = db.Path(pts=pt,width=0)
point_cell.shapes(l1).insert(point)

top_cell = ly.create_cell("TOP")

i1 = db.DCellInstArray(
  cell_index=point_cell.cell_index(),              
  trans=db.DTrans(db.DTrans.M90, 10.0, 5.0), 
  a=db.DVector(0, 2),                    
  b=db.DVector(2, 0),                    
  na=32768 ,                                   
  nb=32768                                     
)
top_cell.insert(i1)

ly.write("kl_big_array.oas")
```

**Reply by sebastian-goeldi** [reply]
Hm, thanks for that example. It seems to stem from the options call to write.

I can make your example fail if I import the standard save options from kfactory:


```python
import klayout.db as db
from kfactory.kcell import save_layout_options

ly = db.Layout()
f_cell = ly.create_cell("F")
l1 = ly.layer(1, 0)

point_cell = ly.create_cell("P")
pts = [db.DPoint(0, 0), db.DPoint(1, 0)]
point = db.DPath(pts=pts, width=0.5)
point_cell.shapes(l1).insert(point)

top_cell = ly.create_cell("TOP")

i1 = db.DCellInstArray(
    cell_index=point_cell.cell_index(),
    trans=db.DTrans(db.DTrans.M90, 10.0, 5.0),
    a=db.DVector(0, 2),
    b=db.DVector(2, 0),
    na=32768,
    nb=32768,
)
top_cell.insert(i1)

print(save_layout_options())
ly.write("kl_big_array.oas", options=save_layout_options())
```

My guess is that if `gds2_[...]` options are set, the gds2 writer still gets called somehow. I'll file an issue with klayout.

**Reply by sebastian-goeldi** [reply, marked as answer]
Ok, investigated more, it's the pure presence of save options, see https://github.com/KLayout/klayout/issues/1952

As a workaround for now, you can use `c._kdb_cell.write("big_array.oas")`. This bypasses kfactory's default write and calls the raw `db.Cell.write`. NB: you **must not** pass the options arg, as it fails otherwise.

**Comment by stefanottili** [comment]
Is this a theoretical exercise or is there an actual use case for using an array with row/col sizes >  32768 ?
And if the latter, what would that use case be ?

**Comment by ghost** [comment]
The use case is for EBL exposures that contain a main 5 x 5 mm array of components that have a pitch in the range 100s of nanometers.

Additionally, there are subarrays that vary parameters. For pattern consistency that relates to exposure consistency, having all the arrays represented the same aids in understanding/mapping variation.

So, while there are ways to circumvent the GDS2 limitation it adds a caveat in the effort to generate descriptive patterns that are concise and consistent.




________________________________

**Comment by stefanottili** [comment]
My background is chip P&R, where I’ve never encountered a 5x5 mm array of 100nm structures.
Any pointers to articles or even better picture of how the end result will look like and what it’s for ?

**Reply by ghost** [reply]
fwiw:

There are a few different parallel efforts, but you could use [Focused ion beam lithography for position-controlled nanowire growth](https://iopscience.iop.org/article/10.1088/1361-6528/acd2e1/meta) as an example. 

NTNU has an Elionix ELS-G100 ([ELS-BODEN 100](https://sts-elionix.com/product/els-g100/)). The EBL will break up arrays into subarrays during writing (subfields relative to the field size). It is accurate to look at this as 16 bit fields, with 65k values (even if there is some overhead that restricts it to 50k). 

This is less of a technical restriction and more of a User eXperience concern. I am trying to build up processes that improves UX, reduces the entry barrier, and accelerates design and analysis.

**Reply by stefanottili** [reply]
Do I understand you correctly that you then forfeit using gds as layout input and expect layout tools that read/write oasis to support array row/col info > uint16_t ?

**Reply by stefanottili** [reply]
> 20nm lines can be written over an entire 500μm field without stitching

So there is no accuracy gained from having a 50k * 50k array spanning 5x5mm, right ?
And besides testing the machine, what would you want to print at 50k * 50k at 100nm spacing with 20 nm accuracy ?

**Reply by ghost** [reply]
Short response: Yes.

You could say that I trade one caveat for another and therefore no net benefit (or worse). With regard to the write oas issue, I don't think it's too much to explore as a solution within gdsfactory. With regard to layout compatibility, your implicit point may be compelling, but documenting the solution and it's dis/advantages is of reasonable value to our users.

**Reply by stefanottili** [reply]
Thanks for your confirmation, but I’m still left puzzled as to what exactly you want to print and how that would look like.

**Comment by stefanottili** [comment]
5x5mm at 100nm pitch = 50k. Assuming one doesn’t need negative row/col, the 16 bit gds fields could be treated as unsigned short, allowing 65k values. Awfully close to what you want.

I’m still missing something here. The e-beam writer that I’m aware of write masks with fractured data in very narrow stripes. What kind of machine are you targeting ?

**Comment by klayoutmatthias** [comment]
Sorry to chime in, but the problem was simply, that KLayout was writing GDS even though the file was called ".oas". The write function needs to be told it has to be OASIS. There is an API method to set the format from the proposed file name. Without that: never trust a file name.

Okay, I know. Should be less confusing.

Matthias

**Reply by joamatab** [reply]
oh, that explains why my write('layout.dxf') was not creating a valid DXF file last week :)

How do we set the export format in klayout python API?

**Comment by tvt173** [comment]
ah, i see... thanks Matthias! the script can be corrected like so:

```python
import gdsfactory as gf

@gf.cell
def big_array(size:tuple=(1,1),
                layer:tuple=(1,0),
                columns:int=3,
                rows:int=3,
                column_pitch:float=3,
                row_pitch:float=3,
                ) -> gf.Component:
    c = gf.Component()
    ref1 = c.add_ref(gf.components.rectangle(size=size, layer=layer),
                    columns = columns,
                    rows = rows,
                    column_pitch = column_pitch,
                    row_pitch = row_pitch)
    return c

c = big_array(columns=32767, rows = 32768)

filename = 'big_array.oas'
options = gf.kdb.SaveLayoutOptions()
options.set_format_from_filename(filename)
c.write(filename, save_options=options)
```

i agree that this is pretty confusing... maybe we should set this option by default when using the `write` function from kfactory/gdsfactory? what do you think @joamatab @sebastian-goeldi ?

![image](https://github.com/user-attachments/assets/30576b12-83b4-48a8-826b-2454ec9ac783)

**Reply by sebastian-goeldi** [reply]
Yes that's my immediate plan. I remember reading about these functions and thought they were obsolete due to autodetection.

**Reply by joamatab** [reply]
Sounds good to me
